openex-api/src/main/java/io/openex/helper/InjectHelper.java

```diff
@@ -30,92 +32,93 @@
+  }
 
-    @Autowired
-    public void setInjectRepository(InjectRepository injectRepository) {
-        this.injectRepository = injectRepository;
-    }
+  @Autowired
+  public void setInjectRepository(InjectRepository injectRepository) {
+    this.injectRepository = injectRepository;
+  }
 
-    @Autowired
-    public void setDryInjectRepository(DryInjectRepository dryInjectRepository) {
-        this.dryInjectRepository = dryInjectRepository;
-    }
+  @Autowired
+  public void setDryInjectRepository(DryInjectRepository dryInjectRepository) {
+    this.dryInjectRepository = dryInjectRepository;
+  }
 
-    private List<Audience> getInjectAudiences(Inject inject) {
-        Exercise exercise = inject.getExercise();
-        return inject.isAllAudiences() ?
-                fromIterable(audienceRepository.findAll(fromExercise(exercise.getId()))) : inject.getAudiences();
-    }
+  private List<Audience> getInjectAudiences(Inject inject) {
+    Exercise exercise = inject.getExercise();
+    return inject.isAllAudiences() ?
+        fromIterable(audienceRepository.findAll(fromExercise(exercise.getId()))) : inject.getAudiences();
+  }
 
-    private Stream<ExecutionContext> getUsersFromInjection(Injection injection) {
-        if (injection instanceof DryInject dryInject) {
-            return dryInject.getRun().getUsers().stream()
-                    .map(user -> new ExecutionContext(openExConfig, user, injection, "Dryrun"));
-        } else if (injection instanceof Inject inject) {
-            List<Audience> audiences = getInjectAudiences(inject);
-            return audiences.stream().filter(Audience::isEnabled)
-                    .flatMap(audience -> audience.getUsers().stream()
-                            .map(user -> new ExecutionContext(openExConfig, user, injection, audience.getName())));
-        }
-        throw new UnsupportedOperationException("Unsupported type of Injection");
+  private Stream<Tuple2<User, String>> getUsersFromInjection(Injection injection) {
+    if (injection instanceof DryInject dryInject) {
+      return dryInject.getRun().getUsers().stream()
+          .map(user -> Tuples.of(user, "Dryrun"));
+    } else if (injection instanceof Inject inject) {
+      List<Audience> audiences = getInjectAudiences(inject);
+      return audiences.stream().filter(Audience::isEnabled)
+          .flatMap(audience -> audience.getUsers().stream()
+              .map(user -> Tuples.of(user, audience.getName())));
     }
+    throw new UnsupportedOperationException("Unsupported type of Injection");
+  }
 
-    private List<ExecutionContext> usersFromInjection(Injection injection) {
-        return getUsersFromInjection(injection)
-                .collect(groupingBy(ExecutionContext::getUser)).entrySet().stream()
-                .map(entry -> new ExecutionContext(openExConfig, entry.getKey(), injection,
-                        entry.getValue().stream().flatMap(ua -> ua.getAudiences().stream()).toList()))
```

Comment:
We had a problem here with the ExecutionContext::getUser that was a ProtectUser and when we gave it to the ExecutionContext we had a type problem as ExecutionContext  want a User
So we remove the ExectutionContext creation in getUsersFromInjection and just pass a tuple of user & audience